packages/auto-id/src/utils.ts

```diff
@@ -1,28 +1,73 @@
+import { compactAddLength } from '@polkadot/util'
 import { ObjectIdentifier } from 'asn1js'
-import { randomBytes } from 'crypto'
 
 /**
- * Encodes a given string representation of an OID into its DER format.
- * This function is specifically used to encode signature algorithm OIDs.
+ * Encodes a given string representation of an OID into its DER format,
+ * and then applies SCALE encoding by prepending the length of the DER-encoded data in compact format.
  *
  * @param oid The string representation of the ObjectIdentifier to be encoded.
- * @returns Uint8Array containing the DER encoded OID along with NULL params of X.509 signature algorithm.
+ * @returns Uint8Array containing the SCALE encoded, DER-encoded OID with appended NULL parameter.
  */
 export function derEncodeSignatureAlgorithmOID(oid: string): Uint8Array {
   const objectIdentifier = new ObjectIdentifier({ value: oid })
-  const berArrayBuffer = objectIdentifier.toBER(false)
+  const oidEncoded = objectIdentifier.toBER(false)
 
   // Typically, in X.509, the algorithm identifier is followed by parameters; for many algorithms, this is just NULL.
   const nullParameter = [0x05, 0x00] // DER encoding for NULL
 
   // Calculate the total length including OID and NULL parameter
-  const totalLength = berArrayBuffer.byteLength + nullParameter.length
+  const totalLength = oidEncoded.byteLength + nullParameter.length
 
-  const sequenceHeader = [0x30, totalLength] // 0x30 is the DER tag for SEQUENCE
+  // Construct the sequence header
+  const sequenceHeader = [0x30] // 0x30 is the DER tag for SEQUENCE
+  if (totalLength < 128) {
+    sequenceHeader.push(totalLength) // Short form length
+  } else {
+    // Long form length encoding
+    const lengthBytes = []
+    let tempLength = totalLength
+    while (tempLength > 0) {
+      lengthBytes.push(tempLength & 0xff)
+      tempLength >>= 8
+    }
+    sequenceHeader.push(0x80 | lengthBytes.length, ...lengthBytes.reverse())
+  }
 
-  return new Uint8Array([...sequenceHeader, ...new Uint8Array(berArrayBuffer), ...nullParameter])
+  // Combine the sequence header, OID, and NULL parameter into one Uint8Array
+  const derSequence = new Uint8Array(
+    sequenceHeader.length + oidEncoded.byteLength + nullParameter.length,
+  )
+  derSequence.set(sequenceHeader, 0)
+  derSequence.set(new Uint8Array(oidEncoded), sequenceHeader.length)
+  derSequence.set(nullParameter, sequenceHeader.length + oidEncoded.byteLength)
```

Comment:
This seems like a quite manual way to handle the encoding. Does the asn1js or other libraries not handle this for you? Did you look into [@peculiar/asn1-schema](https://www.npmjs.com/package/@peculiar/asn1-schema)?

Reply:
As far as I can recall, there was issue in bytes count which on debugging I found the current code as solution as a fix.

> Did you look into [@peculiar/asn1-schema](https://www.npmjs.com/package/@peculiar/asn1-schema)

Yes I had, but didn't found anything relevant.

Reply:
This code looks very fragile. Did you look at the [AlgorithmIdentifier class](https://github.com/PeculiarVentures/asn1-schema/blob/c800e1286d52b460a63f665f90d8e74d6ac69067/packages/x509/src/algorithm_identifier.ts). This looks similar to what was used in [auto-kit](https://github.com/subspace/auto-kit/blob/558666033fa0722a538da1be420c1555eed35e50/auto_identity/utils.py#L7)